scipy/_lib/_uarray/_uarray_dispatch.cxx

```diff
@@ -140,30 +140,30 @@ Using these with PyObject_GetAttr is faster than PyObject_GetAttrString which
 has to create a new python string internally.
  */
 struct {
-  py_ref ua_convert;
-  py_ref ua_domain;
-  py_ref ua_function;
+  PyObject* ua_convert;
+  PyObject* ua_domain;
+  PyObject* ua_function;
 
   bool init() {
-    ua_convert = py_ref::steal(PyUnicode_InternFromString("__ua_convert__"));
+    ua_convert = Py_XNewRef(PyUnicode_InternFromString("__ua_convert__"));
     if (!ua_convert)
       return false;
 
-    ua_domain = py_ref::steal(PyUnicode_InternFromString("__ua_domain__"));
+    ua_domain = Py_XNewRef(PyUnicode_InternFromString("__ua_domain__"));
     if (!ua_domain)
       return false;
 
-    ua_function = py_ref::steal(PyUnicode_InternFromString("__ua_function__"));
+    ua_function = Py_XNewRef(PyUnicode_InternFromString("__ua_function__"));
     if (!ua_function)
       return false;
 
     return true;
   }
 
   void clear() {
-    ua_convert.reset();
-    ua_domain.reset();
-    ua_function.reset();
+    // Py_CLEAR(ua_convert);
+    // Py_CLEAR(ua_domain);
+    // Py_CLEAR(ua_function);
```

Comment:
Why leak this? Is module free called without a valid python context?

Reply:
Yes, see https://github.com/scipy/scipy/pull/21218#discussion_r1695769477 and gh-21214 for context. It sounds like this was always wrong, but for debug builds it probably wasn't too relevant - for free-threading it's easier to hit.

Reply:
https://github.com/scipy/scipy/issues/21214 suggests the problematic case is when `globals_free` is not called, so I don't see how changing `globals_free` is relevant.